File: algorithms/Graphs/Easy/BreadthFirstSearch/VSDBreadthFirstSearch.py
# program to traverse the graph using BFS
# Function takes graph as a input parameters

def BFS(graph):
    queue=[] # Is used to implement the DFS logic
    visted=[] # Is used to store visited vertices
    first = list(graph.keys())[0]
    queue.append(first)
    while queue:
        for vertex in graph[queue[0]]:
            if vertex not in visted and vertex not in queue:
                queue.append(vertex)
        visted.append(queue.pop(0))
    return visted
    

# Test drive code:
graph={
    0:[1,2],
    1:[2],
    2:[0,3],
    3:[3]

}
print("The BFS traversal of the graph is :",BFS(graph))

File: algorithms/Graphs/Easy/BreadthFirstSearch/VSDBreadthFirstSearch.java
import java.util.*;
public class VSDBreadthFirstSearch{
    private int v;
    private LinkedList<Integer> lst[];
    //create a graph
    VSDBreadthFirstSearch(int ver){
        v=ver;
        lst=new LinkedList[v];
        for(int i=0;i<v;i++){
            lst[i]=new LinkedList();
        }
    }
    //add edges to graph
    public void addEdge(int ver,int ele){
        lst[ver].add(ele);
    }
    //BFS algorithm
    public void bfs(int s){
        boolean[] visited=new boolean[v];
        LinkedList<Integer> queue=new LinkedList();
        visited[s]=true;
        queue.add(s);
        while(queue.size()!=0){
            //removing and printing thr removed element
            s=queue.poll();
            System.out.print(s+" ");
            // iterating graph
            Iterator<Integer> i=lst[s].listIterator();
            while(i.hasNext()){
                int n=i.next();
                if(!visited[n]){
                    visited[n]=true;
                    queue.add(n);
                }
            }
        }
    }
    //driver code
    public static void main(String args[]){
        VSDBreadthFirstSearch obj=new VSDBreadthFirstSearch(6);
        obj.addEdge(0, 1);
        obj.addEdge(0, 3);
        obj.addEdge(0, 4);
        obj.addEdge(4, 5);
        obj.addEdge(3, 5);
        obj.addEdge(1, 2);
        obj.addEdge(1, 0);
        obj.addEdge(2, 1);
        obj.addEdge(4, 1);
        obj.addEdge(3, 1);
        obj.addEdge(5, 4);
        obj.addEdge(5, 3);
        System.out.println("The Breadth First Traversal of the graph is as follows(from 0th vertex):");
        obj.bfs(0);
    }
}